### ISSUE TITLE
Saving and loading GraphConv Models

### ISSUE THREAD

**Issue by DarnellGranberry** [deepchem/deepchem#1629]
**Description:**

- DeepChem Version: 2.1.1
- Platform: Windows 10, Python 3.6

Hello, I am trying to train some GraphConvModels and use them in another script to predict properties on a dataset, but I am having difficulty loading the models in the other script.

I tried using model.restore(), which produced the following traceback:
File "C:\Users\darne\Anaconda3\lib\site-packages\deepchem\models\keras_model.py", line 942, in restore
    self._checkpoint.restore(checkpoint).run_restore_ops(self.session)
AttributeError: 'GraphConvModel' object has no attribute '_checkpoint'

I have since tried using model.reload(), which produced the following traceback:
File "C:\Users\darne\Anaconda3\lib\site-packages\deepchem\models\models.py", line 88, in reload
    "Each model is responsible for its own reload method.")
NotImplementedError: Each model is responsible for its own reload method.

In the other issues that I looked at, there was mention of working with the underlying TensorGraph or using Tensorflow's Saver, but have these since been integrated into the back-end of model saving/loading? Either way, is there a reliable way to load these models (and deepchem models in general) from the checkpoints that are created during training? Thanks for your time.

```
### training model, proceeds just fine
from deepchem.utils.save import load_from_disk
dataset_file= "./lipophilicity.csv"
dataset = load_from_disk(dataset_file)

import deepchem as dc
featurizer = dc.feat.ConvMolFeaturizer()
loader = dc.data.CSVLoader(tasks=["exp"], smiles_field="smiles",featurizer=featurizer)
dataset = loader.featurize(dataset_file)

splitter = dc.splits.RandomSplitter(dataset_file)
train_dataset, valid_dataset, test_dataset = splitter.train_valid_test_split(dataset)

transformers = [dc.trans.NormalizationTransformer(transform_y=True, dataset=train_dataset)]

for dataset in [train_dataset, valid_dataset, test_dataset]:
  for transformer in transformers:
      dataset = transformer.transform(dataset)

from deepchem.models.tensorgraph.models.graph_models import GraphConvModel
model = GraphConvModel(n_tasks=1, mode='regression', batch_size=50, model_dir="./models/logD", random_seed=0)

model.fit(train_dataset, nb_epoch=30, deterministic=True)


###Attempt 1 to load in other script:
logD_model = dc.models.GraphConvModel(n_tasks=1, mode='regression', batch_size=50, random_seed=0, model_dir="./models/logD")
logD_model.restore()

###Attempt 2 to load in other script:
logD_model = dc.models.GraphConvModel(n_tasks=1, mode='regression', batch_size=50, random_seed=0, model_dir="./models/logD")
logD_model.reload()
```

**Comment by peastman** [comment]
You said you're using 2.1.1, which is a somewhat old version (the latest release is 2.2), but the stack trace lists classes from the current development code that didn't exist even in 2.2.  How did you install DeepChem?  I'm guessing you used git to check out the repository and install from source?  Can you check what git revision it is?

**Comment by DarnellGranberry** [comment]
Conda lists deepchem as 2.2.1.dev76, but printing the version to the console says it 2.1.1. I did install deepchem from source. The git revision is ff21bdac0f82311cde684e4f081ea5a40472806d.

**Comment by VIGS25** [comment]
`restore` should have the following addition:
```
def restore(self, checkpoint=None):
  self._ensure_built()
```

**Comment by peastman** [comment]
Thanks!  We're in the middle of a major architectural change, replacing TensorGraph with a new class called KerasModel.  That code is still somewhat immature and changing very quickly.

**Comment by DarnellGranberry** [comment]
That fixed the issue. Thanks!

**Comment by DarnellGranberry** [comment]
Hello. Making predictions using the same network (no retraining) on the same dataset returns different predictions. Should this not be deterministic once training is complete? I think there is an error that occurs in saving/loading the model. The code below consistently raises an AssertionError. Is this an error that may also be due to the versioning issue above, or something else? Thanks.


``` import deepchem as dc
import numpy as np

dataset_file= "../data/enamineSubset10K.csv"
featurizer = dc.feat.ConvMolFeaturizer()
bace_loader = dc.data.CSVLoader(tasks=["bace"], smiles_field="SMILES",featurizer=featurizer)
bace_dataset = bace_loader.featurize(dataset_file)

predictions = []
###restore model and make prediction
for _ in range(5):
    bace_model = dc.models.GraphConvModel(n_tasks=1, mode='regression', batch_size=50, random_seed=0, model_dir="../models/bace")
    bace_model.restore()
    bace_predicted = bace_model.predict(bace_dataset)
    predictions.append(bace_predicted)
    print("Prediction complete.")
for i in range( len(predictions)-1 ):
    print(i)
    assert np.array_equal(predictions[i], predictions[i+1])

print("Done.")

**Comment by VIGS25** [comment]
Can you try `np.testing.assert_array_almost_equal(predictions[i], predictions[i+1])` instead of your assert statement?

**Comment by DarnellGranberry** [comment]
This still failed and reported a 100% mismatch. When I plotted the histograms of some of the results before, they could be vastly different (normal dist vs bimodal).

**Comment by peastman** [comment]
I can reproduce this.  Here's a fully self contained script that demonstrates it.

```python
import deepchem as dc
import numpy as np

tasks, datasets, transformers = dc.molnet.load_bace_classification(featurizer='GraphConv')
train_dataset, valid_dataset, test_dataset = datasets
model = dc.models.GraphConvModel(n_tasks=1, mode='regression')
model.fit(train_dataset)

predictions = []
for _ in range(5):
    bace_model = dc.models.GraphConvModel(n_tasks=1, mode='regression', model_dir=model.model_dir)
    bace_model.restore()
    bace_predicted = bace_model.predict(valid_dataset)
    predictions.append(bace_predicted)
for i in range(len(predictions)-1):
    print(i)
    assert np.array_equal(predictions[i], predictions[i+1])
```

This seems to be specific to GraphConvModel.  If instead I use ECFP featurization and a MultitaskRegressor, it passes.  Let me see if I can figure out what's going on.

**Comment by peastman** [comment]
I think this may be the same problem as #1642.  In my testing it fails on TensorFlow 1.13.1, but works correctly on 1.14.  Can you confirm if that's true for you?

**Comment by rbharath** [comment]
I'm going to go ahead and close this issue since I think it was solved in TF 1.14. Please let me know if we should re-open the issue.
